### ISSUE TITLE
How to use cutblur while in classifying images in training loop ?

### ISSUE THREAD

**Issue by IamSparky** [clovaai/cutblur#12]
Here's how I write the training loop ...

`
def train_loop_fn(data_loader, model, optimizer, device, scheduler):
    running_loss = 0.0
    model.train()
    
    for inputs,labels in data_loader:
        inputs = inputs.to(device, dtype=torch.float)
        labels = labels.to(device, dtype=torch.float)

        optimizer.zero_grad()

        outputs = model(inputs)
        loss = loss_fn(outputs, labels)

        loss.backward()
        optimizer_step(optimizer)

        running_loss += loss.item() * inputs.size(0)

    train_loss = running_loss / float(len(train_dataset))
    scheduler.step(train_loss)
    
    print('training Loss: {:.4f}'.format(train_loss))
`

Please tell me how to use cutblur in this loop ?

**Comment by nmhkahn** [comment]
In SR, applying CutBlur is straightforward since we have both LR and HR images.
But in the classification problem, we only have a single image that can think as an HR image so that we need to generate an LR image from an HR one.
I haven't experimented on this but below is one possible scenario.
```python
for inputs,labels in data_loader:
    inputs_HQ = inputs.to(device, dtype=torch.float)
    labels = labels.to(device, dtype=torch.float)

    # or you can apply random noise, jittering, etc..
    inputs_LQ = F.interpolate(inputs_HQ, scale_factor=1/4, mode="bilinear")
    inputs = apply_cutblur(inputs_HQ, inputs_LQ) # just a pseudo-code

    outputs = model(inputs)
```

I'm not sure that the above improves the classification performance but it might improve the robustness of the model.

**Comment by IamSparky** [comment]
I have tried to use cutblur in my training loop but it is showing this error here ; Please help me in fixing it .
![image](https://user-images.githubusercontent.com/42636586/100479846-6d22bf00-3115-11eb-9c98-627187087e8d.png)

# defining the training loop
```
from cutblur.augments import cutblur
def train_loop_fn(data_loader, model, optimizer, device, scheduler=None):
    running_loss = 0.0
    running_corrects = 0
    
    model.train()
    
    for inputs,labels in data_loader:
        inputs_HQ = inputs.to(device, dtype=torch.float)
        labels = labels.to(device, dtype=torch.float)

        # or you can apply random noise, jittering, etc..
        inputs_LQ = F.interpolate(inputs_HQ, scale_factor=1/4, mode="bilinear")
        inputs = cutblur(inputs_HQ, inputs_LQ)
        
        optimizer.zero_grad()

        outputs = model(inputs)
        _, preds = torch.max(outputs, 1)
        
        loss = loss_fn(outputs, label)

        loss.backward()
        xm.optimizer_step(optimizer)

        running_loss += loss.item()
        running_corrects += torch.sum(preds == label.data)
            
    train_loss = running_loss / float(len(train_data))
    train_acc = running_corrects.double() / float(len(train_data))
    
    scheduler.step(train_loss)
    
    xm.master_print('training Loss: {:.4f} & training accuracy : {:.4f}'.format(train_loss , train_acc))

```
Here's the **[link to my notebook](https://www.kaggle.com/soumochatterjee/cutblur-leaf-disease-classification)**

**Comment by nmhkahn** [comment]
add `inputs_LQ = F.interpolate(inputs_LQ, scale_factor=4, mode="bilinear")` right after the downsample line to match the resolution of LR and HR.

**Comment by IamSparky** [comment]
@nmhkahn Which downsample line you are referring [here](https://www.kaggle.com/soumochatterjee/cutblur-leaf-disease-classification) , I am unable to understand. Please help !

**Comment by nmhkahn** [comment]
I was meant to as
```python
from cutblur.augments import cutblur
def train_loop_fn(data_loader, model, optimizer, device, scheduler=None):
    running_loss = 0.0
    running_corrects = 0
    
    model.train()
    
    for inputs,labels in data_loader:
        inputs_HQ = inputs.to(device, dtype=torch.float)
        labels = labels.to(device, dtype=torch.float)

        # or you can apply random noise, jittering, etc..
        inputs_LQ = F.interpolate(inputs_HQ, scale_factor=1/4, mode="bilinear")
        inputs_LQ = F.interpolate(inputs_LQ, scale_factor=4, mode="bilinear") # ADD HERE
        inputs = cutblur(inputs_HQ, inputs_LQ)
```
